packages/react-reconciler/src/ReactFiber.js

```diff
@@ -311,67 +311,99 @@ export function createFiberFromElement(
   const type = element.type;
   const key = element.key;
   const pendingProps = element.props;
-  if (typeof type === 'function') {
-    fiber = shouldConstruct(type)
-      ? createFiber(ClassComponent, pendingProps, key, internalContextTag)
-      : createFiber(
-          IndeterminateComponent,
+
+  switch (type) {
+    case REACT_FRAGMENT_TYPE:
+      return createFiberFromFragment(
+        pendingProps.children,
+        internalContextTag,
+        expirationTime,
+        key,
+      );
+    case REACT_CALL_TYPE:
+      fiber = createFiber(CallComponent, pendingProps, key, internalContextTag);
+      fiber.type = REACT_CALL_TYPE;
+      fiber.expirationTime = expirationTime;
+      return fiber;
+    case REACT_RETURN_TYPE:
+      fiber = createFiber(
+        ReturnComponent,
+        pendingProps,
+        key,
+        internalContextTag,
+      );
+      fiber.type = REACT_RETURN_TYPE;
+      fiber.expirationTime = expirationTime;
+      return fiber;
+    default: {
+      if (typeof type === 'function') {
```

Comment:
We should put these type checks first before the switch. Always make sure that the common path is first.